src/plugins/toml/utility.c

```diff
@@ -0,0 +1,386 @@
+	FILE * f = stdout; // fopen ("keys_mem.txt", "w");
+	for (size_t i = 0; i < size; i++)
+	{
+		Key * key = keys[i];
+		fprintf (f, "KEY = %s, VALUE = %s\n", keyName (key), keyString (key));
+		keyRewindMeta (key);
+		for (const Key * meta = keyNextMeta (key); meta != NULL; meta = keyNextMeta (key))
+		{
+			fprintf (f, "\tMETA KEY = %s, VALUE = %s\n", keyName (meta), keyString (meta));
+		}
+	}
+}
+
+Key * keyAppendIndex (size_t index, const Key * parent)
+{
+	// Key * indexKey = keyDup (parent);
+	Key * indexKey = keyNew (keyName (parent), KEY_END);
+
+	char * indexStr = indexToArrayString (index);
+	keyAddBaseName (indexKey, indexStr);
+	elektraFree (indexStr);
+	return indexKey;
+}
+
+void keyUpdateArrayMetakey (Key * key, size_t newIndex)
+{
+	char * indexStr = indexToArrayString (newIndex);
+	keySetMeta (key, "array", indexStr);
+	elektraFree (indexStr);
+}
+
+char * indexToArrayString (size_t index)
+{
+	size_t digits = 1;
+	for (size_t value = index; value > 9; digits++)
+	{
+		value /= 10;
+	}
+	int strLen = 1 +	    //  '#'
+		     (digits - 1) + // underscores
+		     digits +	    // actual digits
+		     1;		    // '\0'
+	char * str = (char *) elektraCalloc (sizeof (char) * strLen);
+	memset (str, '_', sizeof (char) * strLen);
+	str[0] = '#';
+	str[strLen - 1] = 0;
+	snprintf (str + 1 + (digits - 1), strLen, "%lu", index);
+	return str;
+}
+
+size_t arrayStringToIndex (const char * indexStr)
+{
+	if (*indexStr++ != '#')
+	{
+		return 0;
+	}
+	while (*indexStr == '_')
+	{
+		indexStr++;
+	}
+	size_t val = 0;
+	if (sscanf (indexStr, "%lu", &val) == EOF)
+	{
+		ELEKTRA_ASSERT (0, "Could not parse array index");
+		return 0;
+	}
+	return val;
+}
+
+bool isArrayIndex (const char * basename)
+{
+	if (*basename++ != '#')
+	{
+		return false;
+	}
+	while (*basename == '_')
+	{
+		basename++;
+	}
+	while (*basename >= '0' && *basename <= '9')
+	{
+		basename++;
+	}
+	return *basename == 0;
+}
+
+bool isArrayElement (const Key * key)
+{
+	const char * part = (const char *) keyUnescapedName (key);
+	const char * stop = part + keyGetUnescapedNameSize (key);
+	while (part < stop)
+	{
+		if (isArrayIndex (part))
+		{
+			return true;
+		}
+		part += elektraStrLen (part);
+	}
+	return false;
+}
+
+bool isEmptyArray (Key * key)
+{
+	const Key * meta = findMetaKey (key, "array");
+	ELEKTRA_ASSERT (meta != NULL, "Supplied key must have array meta key, but hadn't");
+	const char * sizeStr = keyString (meta);
+	return elektraStrLen (sizeStr) == 1;
+}
+
+size_t getArrayMax (Key * key)
+{
+	const Key * meta = findMetaKey (key, "array");
+	ELEKTRA_ASSERT (meta != NULL, "Supplied key must have array meta key, but hadn't");
+
+	return arrayStringToIndex (keyString (meta));
+}
+
+const Key * findMetaKey (Key * key, const char * metakeyName)
+{
+	keyRewindMeta (key);
+	for (const Key * meta = keyNextMeta (key); meta != NULL; meta = keyNextMeta (key))
+	{
+		if (elektraStrCmp (keyName (meta), metakeyName) == 0)
+		{
+			return meta;
+		}
+	}
+	return NULL;
+}
```

Comment:
@bauhaus93 Any particular reason, why you didn't just use `keyGetMeta` instead of looping through all metakeys?

Just asking, because it just caused me some confusing while rebasing #3447.

Reply:
No, there wasn't any. I just somehow missed the existence of `keyGetMeta`. I'll replace all `findMetaKey`occurences with `keyGetMeta`, which does exactly what I want. Thanks for the hint!